server/src/main/java/com/scalar/db/server/AdminService.java

```diff
@@ -0,0 +1,73 @@
+package com.scalar.db.server;
+
+import com.google.inject.Inject;
+import com.google.protobuf.Empty;
+import com.scalar.admin.rpc.AdminGrpc;
+import com.scalar.admin.rpc.PauseRequest;
+import com.scalar.admin.rpc.StatsResponse;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.json.Json;
+import javax.json.JsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ThreadSafe
+public class AdminService extends AdminGrpc.AdminImplBase {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AdminService.class);
+
+  private static final long MAX_PAUSE_WAIT_TIME_MILLIS = 10000; // 10 seconds
+
+  private final Pauser pauser;
+
+  @Inject
+  public AdminService(Pauser pauser) {
+    this.pauser = pauser;
+  }
+
+  @Override
+  public void pause(PauseRequest request, StreamObserver<Empty> responseObserver) {
+    pauser.pause();
+
+    if (request.getWaitOutstanding()) {
+      LOGGER.warn("Pausing... waiting until outstanding requests are all finished");
+      long start = System.currentTimeMillis();
+      while (pauser.outstandingRequestsExist()) {
+        try {
+          TimeUnit.MILLISECONDS.sleep(100);
+        } catch (InterruptedException ignored) {
+        }
+
+        if (System.currentTimeMillis() - start >= MAX_PAUSE_WAIT_TIME_MILLIS) {
+          String message = "Failed to pause";
+          LOGGER.warn(message);
+          responseObserver.onError(Status.INTERNAL.withDescription(message).asRuntimeException());
+          return;
+        }
+      }
+    }
+
+    LOGGER.warn("Paused");
+
+    responseObserver.onNext(Empty.getDefaultInstance());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void unpause(Empty request, StreamObserver<Empty> responseObserver) {
+    pauser.unpause();
+    LOGGER.warn("Unpaused");
+    responseObserver.onNext(Empty.getDefaultInstance());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void stats(Empty request, StreamObserver<StatsResponse> responseObserver) {
+    JsonObject json =
+        Json.createObjectBuilder().add("outstanding", pauser.getOutstandingRequestCount()).build();
```

Comment:
Sorry, I might not have fully defined it, but it would be great if you can add `total_succeeded` and `total_failed` in addition to the outstanding. (In Scalar DL, we have `Stats` class and I think we need something like this because we are planning to have more stats.)
https://github.com/scalar-labs/scalar/blob/master/ledger/src/main/java/com/scalar/dl/ledger/server/AdminService.java#L55

Reply:
Yes, I knew Scalar DL has `total_succeeded` and `total_failed`, but I didn't get thier use cases, so I didn't add them to Scalar DB Server because the pause/unpause features don't need them.

And IIRC, `total_succeeded` and `total_failed` don't count all the endpoints in Scalar DL but only the `executeContract` endpoint. That’s why I didn't get their use cases.

What uses `total_succeeded` and `total_failed`?